### DISCUSSION TITLE
Environment Variable from shell overwrites all .env*

### DISCUSSION THREAD

**Discussion by joergrech** [vercel/next.js#38053] [Help]
Maybe this is intentional but why is an environment variable from the shell overwriting variables in .env and .env.local? I have the following situation on a MacBook running macOS BigSur 11.6.7:

```shell
printenv | grep MONGODB_DB
MONGODB_DB=db_shell
```

```shell
cat .env | grep MONGODB_DB
MONGODB_DB=db_env
```

```shell
cat .env.local | grep MONGODB_DB
MONGODB_DB=db_env_local
```

... and when I access the variable in the NextJS app using `process.env.MONGODB_DB` I get `db_shell` - which is not what I want as I have to remove all my environment variables that might collide with variables defines in NextJS.

Btw. the page https://nextjs.org/docs/basic-features/environment-variables does not explicitly mention the overwrite by shell variables (except that what `process.env` is meant to say). It would be nice if it was explicitly mentioned in the [Load Order](https://nextjs.org/docs/basic-features/environment-variables#environment-variable-load-order) section

I can understand that this feature might be interesting when someone needs to inject variables from the outside (e.g., in a docker container) but on a developer machine or a server with multiple applications all shell environment variables (including secrets etc.) are then available within the nextjs app.

If this is intentional - is it possible to disable or clear all variables when running the nextjs app (without removing all variables from the shell)?

**Comment by GabenGar** [comment]
> If this is intentional - is it possible to disable or clear all variables when running the nextjs app (without removing all variables from the shell)?

Yes, by using [`dotenv`](https://github.com/motdotla/dotenv). Be aware using it disables the built-in nextjs behaviour and it's up to you to wrangle the `.env` files from that point.

**Comment by icyJoseph** [comment]
Just for completeness, this is the entire point of shell environmental variables... that they are available to applications. If I make a script:

```js
console.log(process.env);
```

And run it with `node index.js`, I'll see my env vars. Some people use other techniques besides `.env` files to inject variables into their deployment systems, and in general this is an area where WebPack has done a lot of good by enabling amazing handling of `.env` vars, inlining their values, and dead-code elimination for example, but has also hidden away complexity that developers ought to know about.

Here's a person with the opposite issue, the shell env vars where not persisted, https://stackoverflow.com/questions/53801753/how-to-forbid-replacing-process-env-variables-during-compilation-in-webpack.

Normally this is fine, because as long as you don't invoke those variables, they won't be inlined by WebPack. What's more, only those starting with `NEXT_PUBLIC_` are visible to the client.

I'd say in this case, the easiest is to just change the variable name, perhaps to something more meaningful to your app, or by having custom scripts in `package.json`: `"dev": "MONGODB_DB='db_shell' next build" for example.

I think the docs are pretty clear though:

```
[Environment Variable Load Order](https://nextjs.org/docs/basic-features/environment-variables#environment-variable-load-order)
Environment variables are looked up in the following places, in order, stopping once the variable is found.

1. process.env
2. .env.$(NODE_ENV).local
3. .env.local (Not checked when NODE_ENV is test.)
4. .env.$(NODE_ENV)
5. .env


For example, if NODE_ENV is development and you define a variable in both .env.development.local and .env, the value in .env.development.local will be used.
```

**Reply by icyJoseph** [reply]
I realise I wrote this without my morning coffee, so apologies if it sounds dry.
